### DISCUSSION TITLE
[Announcement] IBM z/OS Core v1.4.0 has released

### DISCUSSION THREAD

**Discussion by ddimatos** [ansible-collections/ibm_zos_core#582] [Announcement]
Version 1.4.0
=============

New
---

* ``zos_mount``

    * can manage mount operations for a z/OS UNIX System Services (USS) file
      system data set.

Bug fixes and enhancements
--------------------------

* ``zos_copy``

    * enhanced the force option when `force=true` and the remote file or
      data set `dest` is NOT empty, the `dest` will be deleted and recreated
      with the `src` data set attributes, otherwise it will be recreated
      with the `dest` data set attributes.
    * enhanced to optimize how it captures the permission bits state for
      the `dest`. This change now reviews the source files instead of
      traversing the entire `dest` path.
    * enhanced to support creating a parent directory when it does not exist
      in the `dest` path. Prior to this change, if a parent directory
      anywhere in the path did not exist the task would fail as it was
      stated in documentation.
    * enhanced to support system symbols in PARMLIB. System symbols are
      elements that allow different z/OS® systems to share PARMLIB definitions
      while retaining unique values in those definitions. This was fixed in a
      future release through the use of one of the ZOAU dependency but this
      version of `ibm_zos_core` does not support that dependency version so
      this support was added.
    * fixes a bug that when a directory is copied from the controller to the
      managed node and a mode is set, the mode is applied to the directory
      on the managed node. If the directory being copied contains files and
      mode is set, mode will only be applied to the files being copied not
      the pre-existing files.
    * fixes a bug where options were not defined in the module argument spec
      that will result in error when running `ansible-core` v2.11 and using
      options `force` or `mode`.
    * introduced an updated creation policy referred to as precedence rules
      such that if `dest_data_set` is set, this will take precedence.
      If `dest` is an empty data set, the empty data set will be written with
      the expectation its attributes satisfy the copy. If no precedent rule
      has been exercised, `dest` will be created with the same attributes of `src`.
    * introduced new computation capabilities such that if `dest` is a
      nonexistent data set, the attributes assigned will depend on the type
      of `src`. If `src` is a USS file, `dest` will have a Fixed Block (FB)
      record format and the remaining attributes will be computed. If `src`
      is binary, `dest` will have a Fixed Block (FB) record format with a
      record length of 80, block size of 32760, and the remaining attributes
      will be computed.
    * was enhanced for when `src` is a directory and ends with "/", the
      contents of it will be copied into the root of `dest`. It it doesn't
      end with `/`, the directory itself will be copied.
    * was updated to support the ansible.builtin.ssh connection options; for
      further reference refer to the SSH plugin documentation.
    * was updated to take into account the record length when the source is
      a USS file and the destination is a data set with a record length.
      This is done by inspecting the destination data set attributes and
      using these attributes to create a new data set.
    * was updated with the capabilities to define destination data sets
      from within the module. In the case where you are copying to
      a data set destination that does not exist, you can now do so using
      the new module option `dest_data_set`.
    * fixes a bug that did not create a data set on the specified volume.
    * fixes a bug where a number of attributes were not an option when
      using `dest_data_set`.

* ``zos_fetch``

    * was updated to support the  ansible.builtin.ssh   connection options;
      for further reference refer to the SSH plugin documentation.

* ``zos_job_output``

    * was updated to to include the completion code (CC) for each individual
      job step as part of the ret_code response.
    * fixes a bug that returned all ddname's when a specific ddname was
      provided. Now a specific ddname can be returned and all others ignored.
    * was updated to correct possible truncated responses for the ddname
      content. This would occur for jobs with very large amounts of content
      from a ddname.

* ``zos_job_query``

    * was updated to handle when an invalid job ID or job name is used with
      the module and returns a proper response.
    * was updated to support a 7 digit job number ID for when there are
      greater than 99,999 jobs in the history.

* ``zos_job_submit``

    * was enhanced to check for 'JCL ERROR' when jobs are submitted and
      result in a proper module response.
    * was updated to fail fast when a submitted job fails instead of
      waiting a predetermined time.

* ``zos_operator_action_query``

    * response messages were improved with more diagnostic information in
      the event an error is encountered.

* ``zos_ping``

    * was updated to remove the need for the ``zos_ssh`` connection plugin
      dependency.

* ``zos_mount``

    * fixed option `tag_ccsid` to correctly allow for type int.

* ``zos_operator``

    * enhanced to allow for MVS operator `SET` command, `SET` is equivalent
      to the abbreviated `T` command.

Deprecated Features
-------------------

* ``zos_copy``

    * option sftp_port has been deprecated. To set the SFTP port, use the
      supported options in the ansible.builtin.ssh plugin. Refer to
      the `SSH port <https://docs.ansible.com/ansible/latest/collections/ansible/builtin/ssh_connection.html#parameter-port>`__
      option to configure the port used during the modules SFTP transport.
    * module option model_ds has been removed. The model_ds logic is now
      automatically managed and data sets are either created based on the
      src data set or overridden by the new option destination_dataset.
    * option `dest_dataset` has been deprecated and removed in favor of the
      new option `dest_data_set`.

* ``zos_fetch``

    * option sftp_port has been deprecated. To set the SFTP port, use the
      supported options in the ansible.builtin.ssh plugin. Refer to
      the `SSH port <https://docs.ansible.com/ansible/latest/collections/ansible/builtin/ssh_connection.html#parameter-port>`__
      option to configure the port used during the modules SFTP transport.

* ``zos_ssh``

    * connection plugin has been removed, it is no longer required. You must
      remove all playbook references to connection ibm.ibm_zos_core.zos_ssh.

<hr /><em>This discussion was created from the release <a href='https://github.com/ansible-collections/ibm_zos_core/releases/tag/v1.4.0'>release-v1.4.0</a>.</em>

**Comment by couckearthur** [comment]
@ddimatos is there any news on the release date of the v1.5.0 ? We are running on (E)JES, so we would be happy to see the SDSF dependencies finally gone (ZOAU 1.2.x).

**Reply by ddimatos** [reply]
Hi @couckearthur, that tentative date is February 13th, 2023 , which marks our 3rd Anniversary 🤣, we may try to push it out sooner, that has not been decided  yet. To clarify it would be version 1.5.0-beta.1 and released to Galaxy and eligible for promotion to Automation Hub towards the end of the quarter as 1.5.0.

If you are interested you could build the release branch and give it a try sooner, happy to provide instructions, let me know it's fairly straightforward.

**Reply by ddimatos** [reply]
@couckearthur - you can track the progress for 1.5.0-beta.1 in git issue #433

**Reply by couckearthur** [reply]
Hello Demetri, 
I would be happy to build and test out. So please give me instructions. If you are interested I would also like to share what we are testing out with the Ansible collections at our company. Can I maybe send you an email for that?

**Reply by ddimatos** [reply]
Hello @couckearthur, 
Here is our [general doc and instructions](https://ibm.github.io/z_ansible_collections_doc/installation/installation.html#id3) on how to build any of our collections.

Here is the commands I used to build it:
```
git clone git@github.com:ansible-collections/ibm_zos_core.git
cd ibm_zos_core/
git checkout release-v1.5.0-beta.1
ansible-galaxy collection build
```
I am attaching the build as well. If you find any issues, I suggest a new git issue or even discussion is fine. We do have a couple of issues we are looking into for 1.4.0 zos_copy, surely we will forward port if we find they need to be fixed. 
 
[ibm-ibm_zos_core-1.5.0-beta.1.tar.gz](https://github.com/ansible-collections/ibm_zos_core/files/10233607/ibm-ibm_zos_core-1.5.0-beta.1.tar.gz)

Here is my checksum:
MD5 (ibm-ibm_zos_core-1.5.0-beta.1.tar.gz) = ec271cb270cb5db21368ebe79fc43c47

Installation of the above archive can be managed with:
`ansible-galaxy collection install -f ibm-ibm_zos_core-1.5.0-beta.1.tar.gz`
